File: Suspension matlab/Natural_frequency/Tire_Ring_Structural_Mode/Tire_Ring_Structural_Mode.m
clc;
clear;
scriptInfo = dbstack('-completenames');
scriptDir = fileparts(scriptInfo(1).file);
addpath(scriptDir);
params = load_data(fullfile(scriptDir,'data.csv'));
paramNames = fieldnames(params);
for paramIdx = 1:numel(paramNames)
    eval([paramNames{paramIdx} ' = params.(paramNames{paramIdx});']);
end

%% data

R = tire_R;        % radius [m]
E = tire_E;        % rubber modulus [Pa]
V = tire_V;        % 體積
m = tire_m;        % 直量
rho = m/V;        % density [kg/m^3]
A = tire_A;      % 截面積
I = tire_I;      % 截面二次矩

mode_max = tire_mode_max;

%% 計算
freq = zeros(mode_max,1);

fprintf('Mode   Frequency (Hz)\n');
fprintf('----------------------\n');

for n = 1:mode_max

    if n == 1
        freq(n) = 0; % rigid-body mode
        fprintf('%2d     %.2f\n', n, freq(n));
        continue;
    end

    % classical ring bending scaling (stable form)
    omega = sqrt( (E*I)/(rho*A*R^4) ) * (n^2*(n^2-1));

    freq(n) = omega/(2*pi);

    fprintf('%2d     %.2f\n', n, freq(n));

end
% plot
figure;
plot(1:mode_max, freq, '-o');
xlabel('Mode Number');
ylabel('Frequency (Hz)');
title('Tire Ring Mode Frequencies');
grid on;

File: Suspension matlab/Natural_frequency/Tire_Ring_Structural_Mode/Tire_Ring_Structural_Mode.py
from pathlib import Path
import sys

import matplotlib.pyplot as plt
import numpy as np

ROOT = next(p for p in Path(__file__).resolve().parents if (p / "suspension_utils.py").exists())
sys.path.insert(0, str(ROOT))

from suspension_utils import load_data, show_plots


def main() -> None:
    p = load_data(Path(__file__).with_name("data.csv"))
    radius = p["tire_R"]
    youngs_modulus = p["tire_E"]
    volume = p["tire_V"]
    mass = p["tire_m"]
    area = p["tire_A"]
    inertia = p["tire_I"]
    mode_max = int(p["tire_mode_max"])

    rho = mass / volume
    freq = np.zeros(mode_max)

    print("Mode   Frequency (Hz)")
    print("----------------------")
    for n in range(1, mode_max + 1):
        if n == 1:
            freq[n - 1] = 0.0
        else:
            omega = np.sqrt((youngs_modulus * inertia) / (rho * area * radius**4)) * (n**2 * (n**2 - 1))
            freq[n - 1] = omega / (2 * np.pi)
        print(f"{n:2d}     {freq[n - 1]:.2f}")

    plt.figure()
    plt.plot(np.arange(1, mode_max + 1), freq, "-o")
    plt.xlabel("Mode Number")
    plt.ylabel("Frequency (Hz)")
    plt.title("Tire Ring Mode Frequencies")
    plt.grid(True)
    show_plots()


if __name__ == "__main__":
    main()
